README
#   Copyright 2012-2016 -- Michael Felt
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.

This package automates the building of standard OSF projects into installp
formated files.  On completion the installable fileset is placed in
${BASE}/installp/ppc.
As reference - The template created is stored in ${BASE}/buildaix/ while
temporary files are stored in ${BASE}/.buildaix/
Note: 'make clean' and make 'distclean' are not buildaix aware
and shall not remove these directories. You may remove .buildaix/ at any
time, but be cautious about ./buildaix - there may be additional customization.

SYNTAX buildaix
buildaix flags WITH options
-) add to configure arguments
P) define LPP, export lpp=${LPP} i.e., ipv lpp=$PRODUCT.$PACKAGE
F) define FSET
V) define VRMF, export vrmf=${VRMF}
D) copy proposal (need to copy/paste)
p) prefix
e) eprefix

buildaix flags WITHOUT options
h) help
f) force mode, -e Makefile && make -i distclean, rm -f Makefile config.*
E) eprefix=$prefix/${PRODUCT}/${FILESET}
U) define FORCE_UNSAFE_CONFIGURE=1
B) copy buildaix scripts for customization
T) copy template scripts for customization

* see man buildaix for more details

CONFIGURE and MAKE
If the default ./configure command in buildaix does not succeed
run ./configure manually until you have a successful configure run.
Then execute buildaix as the presence of a Makefile in the project root
will cause the script to skip calling configure

MAKE DESTDIR=${TEMPDIR} install
After make has completed, "make DESTDIR=${TEMPDIR} install" is called
to install the package in a scratch area rather than on the development
system.
If this fails, our recommendation is to use the GNU coreutils install
command. Adjust your PATH variable to be sure it is seen.

MKINSTALLP
The mkinstallp.ksh script works from the following assumptions:
NOTE: any files/directories in ${TEMPDIR}/usr/local are removed
my mkinstallp.ksh

These files are located in the USR part of the installp packaging

/etc is used for configuration - although /var/PROJECT/etc is preffered
/var is used for modifiable data

both /etc and /var are considered to be the ROOT part of the fileset

/opt, /var, /etc, /usr form the PKG.PROGRAM.rte (runtime environment) fileset
except /usr/share/man is put into a seperate fileset package
- aixtools.SOMETHING.man.en_US by default.

PATCHES that might be needed! (old versions of AIX!!)
I have had issues with assigning symbolic links in the packages - default mkinstallp
does not like these and exits with an error. Use following patch to "fix it"

Note: before applying the patch you may want/need to remove /usr/sbin/mkinstallp from
/etc/security/tsd/tsd.dat using
# trustchk -d /usr/sbin/mkinstallp

because the checksum will no longer be correct. I also needed to change the file mode
before applying the patch
# chmod u+w /usr/sbin/mkinstallp


--- /usr/sbin/mkinstallp	2010-06-10 01:04:37.000000000 +0200
+++ ../mkinstallp	2012-09-24 12:34:17.000000000 +0200
@@ -2350,12 +2350,13 @@
        fi
        
        # Make sure we have a valid file
-       if [[ ! -e $ROOT_DIR/${FILE_LINE#[ ]*} ]]; then
+       # or symbolic link -- note -- this is a quick hack for existence of the link, not what it points at!
+       if [[ ! -e $ROOT_DIR/${FILE_LINE#[ ]*} && ! -L $ROOT_DIR/${FILE_LINE#[ ]*} ]]; then
          MESG=`/usr/bin/dspmsg -s 20 cmdinstl_e.cat 87 "0503-881 %s: Invalid filename (%s)\n \
  \tin %s\n" "$CMDNAME" "${FILE_LINE#[ ]*}" "$TEMPLATE_FILE"` 
          print >&2 "$MESG"
          exit 1;
-       fi      
+       fi
  
        # Write out the al file, stripping leading spaces if they appear. Note
        #  that the file paths are relative, so prepend a dot to the filename.


CHANGELOG
buildaix 2.1.5.0:  Sun Jan 30 16:55:41 UTC 2022
- merge all libraries (mkXany)
- retab on many of the sources for better standardization
- verify that packaging does not modify uid.gid of system directories
- Ensure /etc, /opt and /dev are uid.gid root.system when they are part
    part of the package contents.

buildaix 2.1.4.0:  Sun Nov 1 15:59:33 2020 +0100
commit 2318f13d58843b8e03b7735f41895be2cff72421 (tag: v2.1.4.0)
Merge: 0db8b6f f081b7c
Author: Michael Felt <aixtools@users.noreply.github.com>
Date:   Sun Nov 1 15:59:33 2020 +0100

## read the git log for the changes


buildaix 2.1.2.1: 2019-12-17
 * when subdir =="static" go one level deeper for the product and set PROGRAM to static

buildaix 2.1.2.0:
 * find git/product-version as source

buildaix 2.1.1.0: 2017-04-12
 * revert change to prefix - until it is integrated into the mkinstallp program
 ** and the stuff added to deal with creating links (and removing them)
 ** to /opt/bin, /opt/lib, etc..

buildaix 2.1.0.0: 2017-02-14
 * change release number - to parallel the new default prefix /opt/${PROGRAM}
 * modify prefix so that it is /opt/${PROGRAM} to create a unique space

buildaix 2.0.21.0 and 2.0.22.0: 2017-02-14
 * add support for "static" builds
 * add 'adt' for /(opt|usr)/(include|lib)/*.[ha] - so not /opt/lib/{package}*/
 * add two scripts to extract X32 and X64 builds - as precusor to mkXany
 ** or manual changes to default build process

buildaix 2.0.20.0: 2018-01-13
 * improve progress feedback

buildaix 2.0.19.0: 2017-12-05
 * source control now in git
 * drop, or put behind $VERBOSE - most of the feedback currently output (mkinstallp.ksh)
 * export PRODUCT and FILESET so they get picked up by aixinfo and mkinstallp

buildaix 2.0.18.0: 2017-10-05
 * add mkXany (and cplib.ksh) to automate making dual (32- and 64-bit) packages
 * work on mkinstallp.ksh - to support more of what makebff.pl can handle (e.g., .cfgfiles)
 ** note: requires a second call to makebff.pl when 'special' stuff is needed.
 ** what is 'special' may depend on AIX level (AIX 6+ mkinstallp does more that AIX5)
 * This includes moving some core around - hopefully to improve legibility
 * auto_merge and hold_new fail at post_i processing, better before .config processing
 * because the .inventory processing fails (mkinstallp makes a default .inventory)
 * * user_merge is less pretty because it gives a lengthy message at the end.
 * * other: works - in that it gets the MIGSAVE directory defined and copies an original
 * * file, but - it may be simplier to write a ._pre_i and .unpre_i to copy files
 * * to $TMPDIR - so that, during successful install/upgrade .config can use the
 * * files in $TMPDIR

buildaix 2.0.17.0: 2017-09-21
 * added config and unconfig scripts to copy/remove include files stored in 
   /opt/buildaix/include (e.g., zlib.h et al) to /usr/include

buildaix 2.0.16.0: 2017-09-08
 * added an extra character that cannot (easily) be in filename in the inventory

buildaix 2.0.15.0: 2017-06-15
 * fixed an error when using -V and -D - the copy was correct,
   but it could not find the .bff file generated
 * added a flag to number "b" beta packages automatically
   creating a new FIX number based on number after the 'b'

buildaix 2.0.14.0: 2017-02-08
 * fixed bug with vrmf not properly reset when zero length

buildaix 2.0.13.0: 2017-02-08
 * fix - add .config, .pre_i, etc. for other sections!
 * edited CHANGELOG - to shorten long description lines
 * fix bug - when aixtools.zlib was installed bzlib.h was not found.
   Check for all alturnates provided by /opt/buildaix/include
 * add support for supercedes config file - that makebff.pl supports,
   but /usr/sbin/mkinstallp does not

buildaix 2.0.12.0: 2016-11-02
 * fix -D output so that it can be piped into ksh for execution
 * use full pathname for grep and egrep to prevent syntax issues when GNU grep
   is also installed
 * fix aixinfo so that "prj" is never a PRODUCT name
 * personal change for me - when subdir is prj - ignore it - asif == "aixtools"
 * change default machine to pwr5 for xlc
 * default -O setting is now -O3 - will add flag tbd to see optimize level
 * have seperate special CFLAGS by compiler, rather than only xlc
   (that breaks gcc)

buildaix 2.0.11.0: 2016-08-24
 * removed remaining comments for the aborted 64-bit packaging
   - finishing the purge of version 2.0.8.0
 * add a call to ./buildaix/bin/prep_install.ksh, if exists,
   for project processing before mkinstallp packaging begins
 * add -qarch=pwr4 and -qmaxmem=-1 to CFLAGS (assumes xlc as compiler!)
 * fixed bug with -p OPTARG processing
 * changed -e to not depend on ${prefix},
   i.e., not always be a subdir of ${prefix}
 * Modify Fileset Description: to always include ${PRODUCT} ${FILESET}
 * make sure opt/lib/charset.alias has a unique name,
   TBD .post_i processing to set symbolic link

buildaix 2.0.10.1: 2016-06-06
 * make sure .I files are world-readable regadless of umask setting
 * restore a lost addition - namely to add a requisites line such as:
   ">0 { *prereq idsldap.clt32bit62.rte 6.2.0.0 *prereq idsldap.clt32bit63.rte 6.3.0.0 }"
   i.e., to add a group requisite specification of needing one or more
   of a group, e.g., >0 is at least 1 of { ... }

buildaix 2.0.10.0: 2016-05-24
 * corrected a typo in buildaix.ksh - so TARGETDIR is provided everywhere
   in option -D output
 * improved interaction between buildaix and mkinstallp when using buildaix -D

buildaix 2.0.9.0: 2016-04-11
 * corrected a bug where the TARGETDIR was not getting updated soon enough
   (-V argument processing)

buildaix 2.0.8.0: starting 2016-03-04
 * remove 64-bit 'mangling' of the filesets packaging - want autotools
   to modify dirnames, not buildaix scripts

buildaix 2.0.7.0: starting 2016-03-02
 * add ' ' (space) as character to catch (buildaix:filename_cleanup())
 * use ':' to seperate old:new name fo file (for later function to move
   files to original name)
 * TODO : put moved in lpp:root component as a file

opt/buildaix/include/zconf.h
/* zconf.h -- configuration of the zlib compression library
 * Copyright (C) 1995-2013 Jean-loup Gailly.
 * For conditions of distribution and use, see copyright notice in zlib.h
 */

/* @(#) $Id: zconf.h 41 2014-11-04 11:52:30Z michael $ */

#ifndef ZCONF_H
#define ZCONF_H

/*
 * If you *really* need a unique prefix for all types and library functions,
 * compile with -DZ_PREFIX. The "standard" zlib should be compiled without it.
 * Even better than compiling with -DZ_PREFIX would be to use configure to set
 * this permanently in zconf.h using "./configure --zprefix".
 */
#ifdef Z_PREFIX     /* may be set to #if 1 by ./configure */
#  define Z_PREFIX_SET

/* all linked symbols */
#  define _dist_code            z__dist_code
#  define _length_code          z__length_code
#  define _tr_align             z__tr_align
#  define _tr_flush_bits        z__tr_flush_bits
#  define _tr_flush_block       z__tr_flush_block
#  define _tr_init              z__tr_init
#  define _tr_stored_block      z__tr_stored_block
#  define _tr_tally             z__tr_tally
#  define adler32               z_adler32
#  define adler32_combine       z_adler32_combine
#  define adler32_combine64     z_adler32_combine64
#  ifndef Z_SOLO
#    define compress              z_compress
#    define compress2             z_compress2
#    define compressBound         z_compressBound
#  endif
#  define crc32                 z_crc32
#  define crc32_combine         z_crc32_combine
#  define crc32_combine64       z_crc32_combine64
#  define deflate               z_deflate
#  define deflateBound          z_deflateBound
#  define deflateCopy           z_deflateCopy
#  define deflateEnd            z_deflateEnd
#  define deflateInit2_         z_deflateInit2_
#  define deflateInit_          z_deflateInit_
#  define deflateParams         z_deflateParams
#  define deflatePending        z_deflatePending
#  define deflatePrime          z_deflatePrime
#  define deflateReset          z_deflateReset
#  define deflateResetKeep      z_deflateResetKeep
#  define deflateSetDictionary  z_deflateSetDictionary
#  define deflateSetHeader      z_deflateSetHeader
#  define deflateTune           z_deflateTune
#  define deflate_copyright     z_deflate_copyright
#  define get_crc_table         z_get_crc_table
#  ifndef Z_SOLO
#    define gz_error              z_gz_error
#    define gz_intmax             z_gz_intmax
#    define gz_strwinerror        z_gz_strwinerror
#    define gzbuffer              z_gzbuffer
#    define gzclearerr            z_gzclearerr
#    define gzclose               z_gzclose
#    define gzclose_r             z_gzclose_r
#    define gzclose_w             z_gzclose_w
#    define gzdirect              z_gzdirect
#    define gzdopen               z_gzdopen
#    define gzeof                 z_gzeof
#    define gzerror               z_gzerror
#    define gzflush               z_gzflush
#    define gzgetc                z_gzgetc
#    define gzgetc_               z_gzgetc_
#    define gzgets                z_gzgets
#    define gzoffset              z_gzoffset
#    define gzoffset64            z_gzoffset64
#    define gzopen                z_gzopen
#    define gzopen64              z_gzopen64
#    ifdef _WIN32
#      define gzopen_w              z_gzopen_w
#    endif
#    define gzprintf              z_gzprintf
#    define gzvprintf             z_gzvprintf
#    define gzputc                z_gzputc
#    define gzputs                z_gzputs
#    define gzread                z_gzread
#    define gzrewind              z_gzrewind
#    define gzseek                z_gzseek
#    define gzseek64              z_gzseek64
#    define gzsetparams           z_gzsetparams
#    define gztell                z_gztell
#    define gztell64              z_gztell64
#    define gzungetc              z_gzungetc
#    define gzwrite               z_gzwrite
#  endif
#  define inflate               z_inflate
#  define inflateBack           z_inflateBack
#  define inflateBackEnd        z_inflateBackEnd
#  define inflateBackInit_      z_inflateBackInit_
#  define inflateCopy           z_inflateCopy
#  define inflateEnd            z_inflateEnd
#  define inflateGetHeader      z_inflateGetHeader
#  define inflateInit2_         z_inflateInit2_
#  define inflateInit_          z_inflateInit_
#  define inflateMark           z_inflateMark
#  define inflatePrime          z_inflatePrime
#  define inflateReset          z_inflateReset
#  define inflateReset2         z_inflateReset2
#  define inflateSetDictionary  z_inflateSetDictionary
#  define inflateGetDictionary  z_inflateGetDictionary
#  define inflateSync           z_inflateSync
#  define inflateSyncPoint      z_inflateSyncPoint
#  define inflateUndermine      z_inflateUndermine
#  define inflateResetKeep      z_inflateResetKeep
#  define inflate_copyright     z_inflate_copyright
#  define inflate_fast          z_inflate_fast
#  define inflate_table         z_inflate_table
#  ifndef Z_SOLO
#    define uncompress            z_uncompress
#  endif
#  define zError                z_zError
#  ifndef Z_SOLO
#    define zcalloc               z_zcalloc
#    define zcfree                z_zcfree
#  endif
#  define zlibCompileFlags      z_zlibCompileFlags
#  define zlibVersion           z_zlibVersion

/* all zlib typedefs in zlib.h and zconf.h */
#  define Byte                  z_Byte
#  define Bytef                 z_Bytef
#  define alloc_func            z_alloc_func
#  define charf                 z_charf
#  define free_func             z_free_func
#  ifndef Z_SOLO
#    define gzFile                z_gzFile
#  endif
#  define gz_header             z_gz_header
#  define gz_headerp            z_gz_headerp
#  define in_func               z_in_func
#  define intf                  z_intf
#  define out_func              z_out_func
#  define uInt                  z_uInt
#  define uIntf                 z_uIntf
#  define uLong                 z_uLong
#  define uLongf                z_uLongf
#  define voidp                 z_voidp
#  define voidpc                z_voidpc
#  define voidpf                z_voidpf

/* all zlib structs in zlib.h and zconf.h */
#  define gz_header_s           z_gz_header_s
#  define internal_state        z_internal_state

#endif

#if defined(__MSDOS__) && !defined(MSDOS)
#  define MSDOS
#endif
#if (defined(OS_2) || defined(__OS2__)) && !defined(OS2)
#  define OS2
#endif
#if defined(_WINDOWS) && !defined(WINDOWS)
#  define WINDOWS
#endif
#if defined(_WIN32) || defined(_WIN32_WCE) || defined(__WIN32__)
#  ifndef WIN32
#    define WIN32
#  endif
#endif
#if (defined(MSDOS) || defined(OS2) || defined(WINDOWS)) && !defined(WIN32)
#  if !defined(__GNUC__) && !defined(__FLAT__) && !defined(__386__)
#    ifndef SYS16BIT
#      define SYS16BIT
#    endif
#  endif
#endif

/*
 * Compile with -DMAXSEG_64K if the alloc function cannot allocate more
 * than 64k bytes at a time (needed on systems with 16-bit int).
 */
#ifdef SYS16BIT
#  define MAXSEG_64K
#endif
#ifdef MSDOS
#  define UNALIGNED_OK
#endif

#ifdef __STDC_VERSION__
#  ifndef STDC
#    define STDC
#  endif
#  if __STDC_VERSION__ >= 199901L
#    ifndef STDC99
#      define STDC99
#    endif
#  endif
#endif
#if !defined(STDC) && (defined(__STDC__) || defined(__cplusplus))
#  define STDC
#endif
#if !defined(STDC) && (defined(__GNUC__) || defined(__BORLANDC__))
#  define STDC
#endif
#if !defined(STDC) && (defined(MSDOS) || defined(WINDOWS) || defined(WIN32))
#  define STDC
#endif
#if !defined(STDC) && (defined(OS2) || defined(__HOS_AIX__))
#  define STDC
#endif

#if defined(__OS400__) && !defined(STDC)    /* iSeries (formerly AS/400). */
#  define STDC
#endif

#ifndef STDC
#  ifndef const /* cannot use !defined(STDC) && !defined(const) on Mac */
#    define const       /* note: need a more gentle solution here */
#  endif
#endif

#if defined(ZLIB_CONST) && !defined(z_const)
#  define z_const const
#else
#  define z_const
#endif

/* Some Mac compilers merge all .h files incorrectly: */
#if defined(__MWERKS__)||defined(applec)||defined(THINK_C)||defined(__SC__)
#  define NO_DUMMY_DECL
#endif

/* Maximum value for memLevel in deflateInit2 */
#ifndef MAX_MEM_LEVEL
#  ifdef MAXSEG_64K
#    define MAX_MEM_LEVEL 8
#  else
#    define MAX_MEM_LEVEL 9
#  endif
#endif

/* Maximum value for windowBits in deflateInit2 and inflateInit2.
 * WARNING: reducing MAX_WBITS makes minigzip unable to extract .gz files
 * created by gzip. (Files created by minigzip can still be extracted by
 * gzip.)
 */
#ifndef MAX_WBITS
#  define MAX_WBITS   15 /* 32K LZ77 window */
#endif

/* The memory requirements for deflate are (in bytes):
            (1 << (windowBits+2)) +  (1 << (memLevel+9))
 that is: 128K for windowBits=15  +  128K for memLevel = 8  (default values)
 plus a few kilobytes for small objects. For example, if you want to reduce
 the default memory requirements from 256K to 128K, compile with
     make CFLAGS="-O -DMAX_WBITS=14 -DMAX_MEM_LEVEL=7"
 Of course this will generally degrade compression (there's no free lunch).

   The memory requirements for inflate are (in bytes) 1 << windowBits
 that is, 32K for windowBits=15 (default value) plus a few kilobytes
 for small objects.
*/

                        /* Type declarations */

#ifndef OF /* function prototypes */
#  ifdef STDC
#    define OF(args)  args
#  else
#    define OF(args)  ()
#  endif
#endif

#ifndef Z_ARG /* function prototypes for stdarg */
#  if defined(STDC) || defined(Z_HAVE_STDARG_H)
#    define Z_ARG(args)  args
#  else
#    define Z_ARG(args)  ()
#  endif
#endif

/* The following definitions for FAR are needed only for MSDOS mixed
 * model programming (small or medium model with some far allocations).
 * This was tested only with MSC; for other MSDOS compilers you may have
 * to define NO_MEMCPY in zutil.h.  If you don't need the mixed model,
 * just define FAR to be empty.
 */
#ifdef SYS16BIT
#  if defined(M_I86SM) || defined(M_I86MM)
     /* MSC small or medium model */
#    define SMALL_MEDIUM
#    ifdef _MSC_VER
#      define FAR _far
#    else
#      define FAR far
#    endif
#  endif
#  if (defined(__SMALL__) || defined(__MEDIUM__))
     /* Turbo C small or medium model */
#    define SMALL_MEDIUM
#    ifdef __BORLANDC__
#      define FAR _far
#    else
#      define FAR far
#    endif
#  endif
#endif

#if defined(WINDOWS) || defined(WIN32)
   /* If building or using zlib as a DLL, define ZLIB_DLL.
    * This is not mandatory, but it offers a little performance increase.
    */
#  ifdef ZLIB_DLL
#    if defined(WIN32) && (!defined(__BORLANDC__) || (__BORLANDC__ >= 0x500))
#      ifdef ZLIB_INTERNAL
#        define ZEXTERN extern __declspec(dllexport)
#      else
#        define ZEXTERN extern __declspec(dllimport)
#      endif
#    endif
#  endif  /* ZLIB_DLL */
   /* If building or using zlib with the WINAPI/WINAPIV calling convention,
    * define ZLIB_WINAPI.
    * Caution: the standard ZLIB1.DLL is NOT compiled using ZLIB_WINAPI.
    */
#  ifdef ZLIB_WINAPI
#    ifdef FAR
#      undef FAR
#    endif
#    include <windows.h>
     /* No need for _export, use ZLIB.DEF instead. */
     /* For complete Windows compatibility, use WINAPI, not __stdcall. */
#    define ZEXPORT WINAPI
#    ifdef WIN32
#      define ZEXPORTVA WINAPIV
#    else
#      define ZEXPORTVA FAR CDECL
#    endif
#  endif
#endif

#if defined (__BEOS__)
#  ifdef ZLIB_DLL
#    ifdef ZLIB_INTERNAL
#      define ZEXPORT   __declspec(dllexport)
#      define ZEXPORTVA __declspec(dllexport)
#    else
#      define ZEXPORT   __declspec(dllimport)
#      define ZEXPORTVA __declspec(dllimport)
#    endif
#  endif
#endif

#ifndef ZEXTERN
#  define ZEXTERN extern
#endif
#ifndef ZEXPORT
#  define ZEXPORT
#endif
#ifndef ZEXPORTVA
#  define ZEXPORTVA
#endif

#ifndef FAR
#  define FAR
#endif

#if !defined(__MACTYPES__)
typedef unsigned char  Byte;  /* 8 bits */
#endif
typedef unsigned int   uInt;  /* 16 bits or more */
typedef unsigned long  uLong; /* 32 bits or more */

#ifdef SMALL_MEDIUM
   /* Borland C/C++ and some old MSC versions ignore FAR inside typedef */
#  define Bytef Byte FAR
#else
   typedef Byte  FAR Bytef;
#endif
typedef char  FAR charf;
typedef int   FAR intf;
typedef uInt  FAR uIntf;
typedef uLong FAR uLongf;

#ifdef STDC
   typedef void const *voidpc;
   typedef void FAR   *voidpf;
   typedef void       *voidp;
#else
   typedef Byte const *voidpc;
   typedef Byte FAR   *voidpf;
   typedef Byte       *voidp;
#endif

#if !defined(Z_U4) && !defined(Z_SOLO) && defined(STDC)
#  include <limits.h>
#  if (UINT_MAX == 0xffffffffUL)
#    define Z_U4 unsigned
#  elif (ULONG_MAX == 0xffffffffUL)
#    define Z_U4 unsigned long
#  elif (USHRT_MAX == 0xffffffffUL)
#    define Z_U4 unsigned short
#  endif
#endif

#ifdef Z_U4
   typedef Z_U4 z_crc_t;
#else
   typedef unsigned long z_crc_t;
#endif

#if 1    /* was set to #if 1 by ./configure */
#  define Z_HAVE_UNISTD_H
#endif

#if 1    /* was set to #if 1 by ./configure */
#  define Z_HAVE_STDARG_H
#endif

#ifdef STDC
#  ifndef Z_SOLO
#    include <sys/types.h>      /* for off_t */
#  endif
#endif

#if defined(STDC) || defined(Z_HAVE_STDARG_H)
#  ifndef Z_SOLO
#    include <stdarg.h>         /* for va_list */
#  endif
#endif

#ifdef _WIN32
#  ifndef Z_SOLO
#    include <stddef.h>         /* for wchar_t */
#  endif
#endif

/* a little trick to accommodate both "#define _LARGEFILE64_SOURCE" and
 * "#define _LARGEFILE64_SOURCE 1" as requesting 64-bit operations, (even
 * though the former does not conform to the LFS document), but considering
 * both "#undef _LARGEFILE64_SOURCE" and "#define _LARGEFILE64_SOURCE 0" as
 * equivalently requesting no 64-bit operations
 */
#if defined(_LARGEFILE64_SOURCE) && -_LARGEFILE64_SOURCE - -1 == 1
#  undef _LARGEFILE64_SOURCE
#endif

#if defined(__WATCOMC__) && !defined(Z_HAVE_UNISTD_H)
#  define Z_HAVE_UNISTD_H
#endif
#ifndef Z_SOLO
#  if defined(Z_HAVE_UNISTD_H) || defined(_LARGEFILE64_SOURCE)
#    include <unistd.h>         /* for SEEK_*, off_t, and _LFS64_LARGEFILE */
#    ifdef VMS
#      include <unixio.h>       /* for off_t */
#    endif
#    ifndef z_off_t
#      define z_off_t off_t
#    endif
#  endif
#endif

#if defined(_LFS64_LARGEFILE) && _LFS64_LARGEFILE-0
#  define Z_LFS64
#endif

#if defined(_LARGEFILE64_SOURCE) && defined(Z_LFS64)
#  define Z_LARGE64
#endif

#if defined(_FILE_OFFSET_BITS) && _FILE_OFFSET_BITS-0 == 64 && defined(Z_LFS64)
#  define Z_WANT64
#endif

#if !defined(SEEK_SET) && !defined(Z_SOLO)
#  define SEEK_SET        0       /* Seek from beginning of file.  */
#  define SEEK_CUR        1       /* Seek from current position.  */
#  define SEEK_END        2       /* Set file pointer to EOF plus "offset" */
#endif

#ifndef z_off_t
#  define z_off_t long
#endif

#if !defined(_WIN32) && defined(Z_LARGE64)
#  define z_off64_t off64_t
#else
#  if defined(_WIN32) && !defined(__GNUC__) && !defined(Z_SOLO)
#    define z_off64_t __int64
#  else
#    define z_off64_t z_off_t
#  endif
#endif

/* MVS linker does not support external names larger than 8 bytes */
#if defined(__MVS__)
  #pragma map(deflateInit_,"DEIN")
  #pragma map(deflateInit2_,"DEIN2")
  #pragma map(deflateEnd,"DEEND")
  #pragma map(deflateBound,"DEBND")
  #pragma map(inflateInit_,"ININ")
  #pragma map(inflateInit2_,"ININ2")
  #pragma map(inflateEnd,"INEND")
  #pragma map(inflateSync,"INSY")
  #pragma map(inflateSetDictionary,"INSEDI")
  #pragma map(compressBound,"CMBND")
  #pragma map(inflate_table,"INTABL")
  #pragma map(inflate_fast,"INFA")
  #pragma map(inflate_copyright,"INCOPY")
#endif

#endif /* ZCONF_H */
